### ISSUE TITLE
Public key recovery from signature fails immediately after signing

### ISSUE THREAD

**Issue by shamatar** [bitcoin-core/secp256k1#497]
Good evening.

I've created iOS build of this library at https://github.com/shamatar/secp256k1_ios, basically it just build it with the following flags

#define USE_NUM_NONE 1
#define USE_FIELD_INV_BUILTIN 1
#define USE_SCALAR_INV_BUILTIN 1
#define ENABLE_MODULE_RECOVERY 1
#define ENABLE_MODULE_ECDH 1
#ifdef __LP64__
typedef __int128 int128_t;
typedef unsigned __int128 uint128_t;
#define USE_FIELD_5X52 1
#define USE_SCALAR_4X64 1
#else
#define USE_FIELD_10X26 1
#define USE_SCALAR_8X32 1
#endif

Although during the usage I encounter an error in the procedure Recoverable signature -> Recover public key from the signature. I supply signature procedure with extra entropy (last parameter)
 and use rfc6979 nonce function
secp256k1_ecdsa_sign_recoverable(..., secp256k1_nonce_function_rfc6979, extraEntropyPointer)

and then immediately check that public key from recovery 
secp256k1_ecdsa_recoverable_signature_parse_compact

and by direct conversion from private key
secp256k1_ec_pubkey_create

of course all public keys are further converted to the same serialized format, although comparison of such keys sometimes fails and I have to retry all this procedure with another extra entropy. Usually number of such rounds is 1 to 5. I wonder if it's expected behavior or I should try to dig somewhere is sources.

I've also tried to use secp256k1_ecdsa_sign_recoverable with no additional entropy and in this case procedure never succeeds for some input parameters.

**Comment by sipa** [comment]
It's hard to follow what exact functions you're using. Can you provide a code snippet that fails?

Do the tests compile and run on your target?

**Comment by shamatar** [comment]
Here are 3 Swift bindings that I use for signing, recovering and converting private to public

```
static func signForRecovery(hash: Data, privateKey: Data) -> (compressed:Data?, uncompressed: Data?) {
        if (hash.count != 32 || privateKey.count != 32) {return (nil, nil)}
        if !SECP256K1.verifyPrivateKey(privateKey: privateKey) {
            return (nil, nil)
        }
        for rounds in 0...1024 {
            var extraEntropy = Data(count: 32)
            for _ in 0...1024 {
                let result = extraEntropy.withUnsafeMutableBytes {
                    SecRandomCopyBytes(kSecRandomDefault, extraEntropy.count, $0)
                }
                if result == errSecSuccess {
                    break
                }
            }
            var uncompressed: secp256k1_ecdsa_recoverable_signature = secp256k1_ecdsa_recoverable_signature();
            var result = hash.withUnsafeBytes { (hashPointer:UnsafePointer<UInt8>) -> Int32 in
                privateKey.withUnsafeBytes { (privateKeyPointer:UnsafePointer<UInt8>) -> Int32 in
                    extraEntropy.withUnsafeBytes { (extraEntropyPointer:UnsafePointer<UInt8>) -> Int32 in
                        let res = secp256k1_ecdsa_sign_recoverable(context!, UnsafeMutablePointer<secp256k1_ecdsa_recoverable_signature>(&uncompressed), hashPointer, privateKeyPointer, secp256k1_nonce_function_rfc6979, extraEntropyPointer)
                        return res
                    }
                }
            }
            if result == 0 {
                continue
//                return (nil, nil)
            }
            let uncompressedSignature = Data(toByteArray(uncompressed))
            var compressedSignature = Data(count: 64)
            var v: Int32 = 0
            result = compressedSignature.withUnsafeMutableBytes { (compressedPointer:UnsafeMutablePointer<UInt8>) -> Int32 in
                let res = secp256k1_ecdsa_recoverable_signature_serialize_compact(context!, compressedPointer, UnsafeMutablePointer<Int32>(&v), UnsafeMutablePointer<secp256k1_ecdsa_recoverable_signature>(&uncompressed))
                return res
            }
            if result == 0 {
                continue
//                return (nil, nil)
            }
            let buffer = toByteArray(v)
            compressedSignature.append(buffer.last!)
            
//            let S = BigUInt(compressedSignature[32..<65])
//            if S > secp256k1_halfN {
//                continue
//            }

            guard let recoveredPubKey = SECP256K1.recoverPublicKey(hash: hash, signature: compressedSignature) else {continue}
            guard let truePubKey = SECP256K1.privateToPublic(privateKey: privateKey) else {continue}
            if recoveredPubKey != truePubKey {continue}
            print("Signature required \(rounds) rounds")
            return (compressedSignature, uncompressedSignature)
        }
        return (nil, nil)
    }
    
    static func privateToPublic(privateKey: Data, compressed: Bool = false) -> Data? {
        if (privateKey.count != 32) {return nil}
        var publicKey = secp256k1_pubkey()
        var result = privateKey.withUnsafeBytes { (privateKeyPointer:UnsafePointer<UInt8>) -> Int32 in
            let res = secp256k1_ec_pubkey_create(context!, UnsafeMutablePointer<secp256k1_pubkey>(&publicKey), privateKeyPointer)
            return res
        }
        if result == 0 {
            return nil
        }
        var keyLength = compressed ? 33 : 65
        var serializedPubkey = Data(count: keyLength)
        result = serializedPubkey.withUnsafeMutableBytes { (serializedPubkeyPointer:UnsafeMutablePointer<UInt8>) -> Int32 in
            let res = secp256k1_ec_pubkey_serialize(context!,
                                                    serializedPubkeyPointer,
                                                    UnsafeMutablePointer<Int>(&keyLength),
                                                    UnsafeMutablePointer<secp256k1_pubkey>(&publicKey),
                                                    UInt32(compressed ? SECP256K1_EC_COMPRESSED : SECP256K1_EC_UNCOMPRESSED))
            return res
        }
        return serializedPubkey
    }
    
    static func recoverPublicKey(hash: Data, signature: Data, compressed: Bool = false) -> Data? {
        guard hash.count == 32, signature.count == 65 else {return nil}
        var uncompressed: secp256k1_ecdsa_recoverable_signature = secp256k1_ecdsa_recoverable_signature();
        let compressedSignature = Data(signature[0..<64])
        let v = Int32(signature[64])
        var result = compressedSignature.withUnsafeBytes{ (compressedSignaturePointer: UnsafePointer<UInt8>) -> Int32 in
                    let res = secp256k1_ecdsa_recoverable_signature_parse_compact(context!, UnsafeMutablePointer<secp256k1_ecdsa_recoverable_signature>(&uncompressed), compressedSignaturePointer, v)
                    return res
                }
        if result == 0 {
            return nil
        }
        var publicKey: secp256k1_pubkey = secp256k1_pubkey()
        result = hash.withUnsafeBytes { (hashPointer:UnsafePointer<UInt8>) -> Int32 in
            withUnsafePointer(to: &uncompressed) { (signaturePointer:UnsafePointer<secp256k1_ecdsa_recoverable_signature>) -> Int32 in
                let res = secp256k1_ecdsa_recover(context!, UnsafeMutablePointer<secp256k1_pubkey>(&publicKey),
                    signaturePointer, hashPointer)
                return res
            }
        }
        if result == 0 {
            return nil
        }
        var keyLength = compressed ? 33 : 65
        var serializedPubkey = Data(count: keyLength)
        result = serializedPubkey.withUnsafeMutableBytes { (serializedPubkeyPointer:UnsafeMutablePointer<UInt8>) -> Int32 in
            let res = secp256k1_ec_pubkey_serialize(context!,
                                                    serializedPubkeyPointer,
                                                    UnsafeMutablePointer<Int>(&keyLength),
                                                    UnsafeMutablePointer<secp256k1_pubkey>(&publicKey),
                                                    UInt32(compressed ? SECP256K1_EC_COMPRESSED : SECP256K1_EC_UNCOMPRESSED))
            return res
        }
        return serializedPubkey
    }
```

I didn't run tests yet, just packaged latest sources as CocoaPod with flags shown above (most general ones as I understand).

Sincerely, Alex

**Comment by sipa** [comment]
Well please make sure the tests run first. Perhaps there is a bug in libsecp256k1 that only affects your build target.

**Comment by shamatar** [comment]
I've built it for macOS using the same parameters as used for iOS:

compiler = clang
./configure --host="none-apple-darwin" --enable-module-recovery --with-asm=no --with-bignum=no 

```
## ----------- ##
## confdefs.h. ##
## ----------- ##

/* confdefs.h */
#define PACKAGE_NAME "libsecp256k1"
#define PACKAGE_TARNAME "libsecp256k1"
#define PACKAGE_VERSION "0.1"
#define PACKAGE_STRING "libsecp256k1 0.1"
#define PACKAGE_BUGREPORT ""
#define PACKAGE_URL ""
#define PACKAGE "libsecp256k1"
#define VERSION "0.1"
#define STDC_HEADERS 1
#define HAVE_SYS_TYPES_H 1
#define HAVE_SYS_STAT_H 1
#define HAVE_STDLIB_H 1
#define HAVE_STRING_H 1
#define HAVE_MEMORY_H 1
#define HAVE_STRINGS_H 1
#define HAVE_INTTYPES_H 1
#define HAVE_STDINT_H 1
#define HAVE_UNISTD_H 1
#define HAVE_DLFCN_H 1
#define LT_OBJDIR ".libs/"
#define HAVE___INT128 1
#define HAVE_BUILTIN_EXPECT 1
#define USE_FIELD_5X52 1
#define USE_NUM_NONE 1
#define USE_FIELD_INV_BUILTIN 1
#define USE_SCALAR_INV_BUILTIN 1
#define USE_SCALAR_4X64 1
#define USE_ECMULT_STATIC_PRECOMPUTATION 1
#define ENABLE_MODULE_RECOVERY 1
```

Tests has passed, I will try to do the same on simulator and real device. 

Also I'd like to check if generation of ecmult_static_context.h on target machine is legitimate, so this header can be just included in the bundle, instead of being generated on target machine.

**Comment by shamatar** [comment]
I've run tests on simulator and the device (64 arch though) and didn't get any errors. With new config parameters signature now requires one attempt. Would you still explain a little about ecmult_static_context.h generation on one machine and using on other.

Sincerely, Alex

**Comment by shamatar** [comment]
Solved, found a binding mistake
